packages/block-editor/src/components/use-block-drop-zone/index.native.js

```diff
@@ -111,50 +116,82 @@ export default function useBlockDropZone( {
 	rootClientId: targetRootClientId = '',
 } = {} ) {
 	const targetBlockIndex = useSharedValue( null );
+	const dragPosition = {
+		x: useSharedValue( 0 ),
+		y: useSharedValue( 0 ),
+	};
+	const prevDragPosition = {
+		x: useSharedValue( 0 ),
+		y: useSharedValue( 0 ),
+	};
 
 	const { getBlockListSettings, getSettings } = useSelect( blockEditorStore );
 	const { blocksLayouts, getBlockLayoutsOrderedByYCoord } =
 		useBlockListContext();
 
 	const getSortedBlocksLayouts = useCallback( () => {
 		return getBlockLayoutsOrderedByYCoord( blocksLayouts.current );
+		// We use the value of `blocksLayouts` as the dependency.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
 	}, [ blocksLayouts.current ] );
 
 	const isRTL = getSettings().isRTL;
 
 	const onBlockDrop = useOnBlockDrop();
 
-	const throttled = useThrottle(
-		useCallback(
-			( event ) => {
-				const sortedBlockLayouts = getSortedBlocksLayouts();
-
-				const targetIndex = getNearestBlockIndex(
-					sortedBlockLayouts,
-					{ x: event.x, y: event.y },
-					getBlockListSettings( targetRootClientId )?.orientation,
-					isRTL
-				);
-				if ( targetIndex !== null ) {
-					targetBlockIndex.value = targetIndex ?? 0;
-				}
-			},
-			[
-				getSortedBlocksLayouts,
-				getNearestBlockIndex,
-				getBlockListSettings,
-				targetBlockIndex,
-			]
-		),
-		200
+	const updateTargetBlockIndex = useCallback(
+		( event ) => {
+			const sortedBlockLayouts = getSortedBlocksLayouts();
+
+			const targetIndex = getNearestBlockIndex(
+				sortedBlockLayouts,
+				{ x: event.x, y: event.y },
+				getBlockListSettings( targetRootClientId )?.orientation,
+				isRTL
+			);
+			if ( targetIndex !== null ) {
+				targetBlockIndex.value = targetIndex ?? 0;
+			}
+		},
+		[
+			getSortedBlocksLayouts,
+			getBlockListSettings,
+			targetRootClientId,
+			isRTL,
+			targetBlockIndex,
+		]
 	);
 
+	useDerivedValue( () => {
+		const x = dragPosition.x.value;
+		const y = dragPosition.y.value;
+		const prevX = prevDragPosition.x.value;
+		const prevY = prevDragPosition.y.value;
+		// `updateTargetBlockIndex` performs expensive calculations, so we throttle
+		// the call using a offset threshold based on the dragging position.
+		if (
+			Math.abs( x - prevX ) >= UPDATE_TARGET_BLOCK_INDEX_THRESHOLD ||
+			Math.abs( y - prevY ) >= UPDATE_TARGET_BLOCK_INDEX_THRESHOLD
+		) {
+			runOnJS( updateTargetBlockIndex )( { x, y } );
+			prevDragPosition.x.value = x;
+			prevDragPosition.y.value = y;
+			return true;
+		}
+		return false;
+	} );
+
 	return {
-		onBlockDragOver( event ) {
-			throttled( event );
+		onBlockDragOver( { x, y } ) {
+			dragPosition.x.value = x;
+			dragPosition.y.value = y;
+		},
```

Comment:
This `onBlockDragOver` does not appeared to be utilized anywhere. Can we remove this?

Reply:
True, it's not being used. I simply added it for convention, so the hook provides two ways to call the handler:
* In the JS thread  with `onBlockDragOver`
* In the UI thread via Reanimated worklets with `onBlockDragOverWorklet`.

If we rather not include unused code I'll be happy to remove it 🙂.

Reply:
I see. If this is an established convention, then retaining the code makes sense. 👍🏻

Reply:
> If this is an established convention, then retaining the code makes sense. 👍🏻

I don't think this is an established convention, or at least I haven't seen it before in the project. But we could take this opportunity as a proposal if that makes sense.

In this case, what I wanted to highlight adding the `Worklet` suffix was mainly to differentiate that the handler is a worklet compared to the rest of the handlers provided by the hook. This way makes the user aware of the two calling options of the handler.

Reply:
> My question would be: does the knowledge of the worklet have an impact on the callback consumer?

Yes, it's important because it impacts how it's called, let me illustrate with an example:
```
const onBlockDragOver = () => {
	// ...
};

const onBlockDragOverWorklet = () => {
	'worklet';
	// ...
};

const func = () => {
	// From the JS thread, we can only call regular functions.
	onBlockDragOver(); // OK ✅
	onBlockDragOverWorklet(); // Fails ❌
	runOnUI( onBlockDragOverWorklet )(); // OK ✅ - This is how we call worklets from the JS thread.
};

const worklet = () => {
	'worklet';
	// When in a worklet, we are in the UI thread, so we can only call other worklets.
	onBlockDragOver(); // Fails ❌
	runOnJS( onBlockDragOver )(); // OK ✅ - This is how we call regular functions from the UI thread.
	onBlockDragOverWorklet(); // OK ✅
};

```